### PR TITLE
clickhouse-client editor

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Now clickhouse-client supports opening EDITOR to edit commands. `Alt-Shift-E` 

Detailed description / Documentation draft:

None

### CODE REVIEW

File: base/common/ReplxxLineReader.cpp [hunk 1]
```diff
@@ -81,6 +87,8 @@ ReplxxLineReader::ReplxxLineReader(
     /// it also binded to M-p/M-n).
     rx.bind_key(Replxx::KEY::meta('N'), [this](char32_t code) { return rx.invoke(Replxx::ACTION::COMPLETE_NEXT, code); });
     rx.bind_key(Replxx::KEY::meta('P'), [this](char32_t code) { return rx.invoke(Replxx::ACTION::COMPLETE_PREVIOUS, code); });
+
+    rx.bind_key(Replxx::KEY::meta('E'), [this](char32_t) { openEditor(); return Replxx::ACTION_RESULT::CONTINUE; });
```
Comment (azat) [inline base/common/ReplxxLineReader.cpp:91]: How about adding default `"\C-x\C-e"` (as readline has for `edit-and-execute-command`) ?

File: base/common/ReplxxLineReader.cpp [hunk 2]
```diff
@@ -125,7 +133,73 @@ void ReplxxLineReader::addToHistory(const String & line)
...
+        std::string str;
+        t.seekg(0, std::ios::end);
+        str.reserve(t.tellg());
+        t.seekg(0, std::ios::beg);
+        str.assign((std::istreambuf_iterator<char>(t)), std::istreambuf_iterator<char>());
+        rx.set_state(replxx::Replxx::State(str.c_str(), str.size()));
+    }
+    if (bracketed_paste_enabled)
+        enableBracketedPaste();
+    ::unlink(filename);
```
Comment (azat) [inline base/common/ReplxxLineReader.cpp:198]: Worth checking return value?

File: base/common/ReplxxLineReader.cpp [hunk 3]
```diff
@@ -125,7 +133,73 @@ void ReplxxLineReader::addToHistory(const String & line)
...
+    char filename[] = "clickhouse_replxx.XXXXXX";
+    int fd = ::mkstemp(filename);
+    if (-1 == fd)
+        return;
+    String editor = std::getenv("EDITOR");
+    if (editor.empty())
+        editor = "vim";
+
+    replxx::Replxx::State state(rx.get_state());
+    write(fd, state.text(), strlen(state.text()));
```
Comment (azat) [inline base/common/ReplxxLineReader.cpp:182]: What do you think about sending formatted query into the editor? Maybe under some option

**Review by azat — COMMENTED** [review]
Nice stuff! (I've though about this recently, but did not bother enough to add this)

File: base/common/ReplxxLineReader.cpp (hunk 1 above)
Comment (amosbird) [reply base/common/ReplxxLineReader.cpp:91]: Sounds reasonable. I wish we can have keybinds exported to some configuration file.

File: base/common/ReplxxLineReader.cpp (hunk 2 above)
Comment (amosbird) [reply base/common/ReplxxLineReader.cpp:198]: But there is no way to recover, and I don't think users would be irritated by leaving some tmp files around.

File: base/common/ReplxxLineReader.cpp (hunk 3 above)
Comment (amosbird) [reply base/common/ReplxxLineReader.cpp:182]: It should be the job of EDITOR, and we don't have clickhouse parser availble here.

**Comment by amosbird** [comment]
P.S. It's worth mentioning that the editor feature is better suit with `-n`, that is, single line editing with multiple queries (bracketed_paste enabled)

File: base/common/ReplxxLineReader.cpp (hunk 2 above)
Comment (azat) [reply base/common/ReplxxLineReader.cpp:213]: >But there is no way to recover

At least `rx.print`

> and I don't think users would be irritated by leaving some tmp files around.

There just should zero of them

File: base/common/ReplxxLineReader.cpp [hunk 4]
```diff
@@ -125,7 +133,106 @@ void ReplxxLineReader::addToHistory(const String & line)
...
+    {
+        rx.print("Cannot waitpid: %s\n", errnoToString(errno).c_str());
+        return -1;
+    }
+    return status;
+}
+
+void ReplxxLineReader::openEditor()
+{
+    char filename[] = "clickhouse_replxx.XXXXXX";
```
Comment (azat) [inline base/common/ReplxxLineReader.cpp:182]: BTW worth adding an `.sql` extension (to allow editors to enable syntax highlight)

Comment (amosbird) [reply base/common/ReplxxLineReader.cpp:182]: but the last six characters of `mkstemp` template must be "XXXXXX"

Comment (amosbird) [reply base/common/ReplxxLineReader.cpp:182]: Agree.

**Comment by amosbird** [comment]
Tests seem unrelated.

**Review by alexey-milovidov — COMMENTED** [review]
Ok. But we need to make a test with `expect`.

**Comment by amosbird** [comment]
> Ok. But we need to make a test with `expect`.

done

File: tests/queries/0_stateless/01610_client_spawn_editor.sh [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+#\
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01610_client_spawn_editor.sh:3]: Do I understand correctly, this file is a source code in both `bash` and `expect`?

Comment (amosbird) [reply tests/queries/0_stateless/01610_client_spawn_editor.sh:3]: Yes...  And it still doesn't work in CI. It works fine on my local box.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01610_client_spawn_editor.sh:3]: Let's rewrite the test without this trick... maybe it will help.

**Comment by filimonov** [comment]
"One improvement would be to open the editor with the current query like MySQL does"

See https://dev.mysql.com/doc/refman/8.0/en/mysql-commands.html

**Pull request merged**